src/include/distributed/version_compat.h

```diff
@@ -240,5 +244,132 @@ RangeVarGetRelidInternal(const RangeVar *relation, LOCKMODE lockmode, uint32 fla
+#define GetSysCacheOid4Compat GetSysCacheOid4
+#define make_aggCompat make_agg
+
+#define fcSetArg(fc, n, argval) \
+	(((fc)->args[n].isnull = false), ((fc)->args[n].value = (argval)))
+	#define fcSetArgNull(fc, n) \
+	(((fc)->args[n].isnull = true), ((fc)->args[n].value = (Datum) 0))
+
+typedef struct
+{
+	File fd;
+	off_t offset;
+} FileCompat;
+
+static inline int
+FileWriteCompat(FileCompat *file, char *buffer, int amount, uint32 wait_event_info)
+{
+	int count = FileWrite(file->fd, buffer, amount, file->offset, wait_event_info);
+	if (count > 0)
+	{
+		file->offset += count;
+	}
+	return count;
+}
+
+
+static inline int
+FileReadCompat(FileCompat *file, char *buffer, int amount, uint32 wait_event_info)
+{
+	int count = FileRead(file->fd, buffer, amount, file->offset, wait_event_info);
+	if (count > 0)
+	{
+		file->offset += count;
+	}
+	return count;
+}
+
+
+static inline FileCompat
+FileCompatFromFileStart(File fileDesc)
+{
+	FileCompat fc = {
+		.fd = fileDesc,
+		.offset = 0
+	};
+
+	return fc;
+}
+
+
+#else /* pre PG12 */
+#define QTW_EXAMINE_RTES_BEFORE QTW_EXAMINE_RTES
+#define MakeSingleTupleTableSlotCompat(tupleDesc, tts_opts) \
+	MakeSingleTupleTableSlot(tupleDesc)
+#define NextCopyFromCompat(cstate, econtext, values, nulls) \
+	NextCopyFrom(cstate, econtext, values, nulls, NULL)
+
+/*
+ * In PG12 GetSysCacheOid requires an oid column,
+ * whereas beforehand the oid column was implicit with WITH OIDS
+ */
+#define GetSysCacheOid1Compat(cacheId, oidcol, key1) \
+	GetSysCacheOid1(cacheId, key1)
+#define GetSysCacheOid2Compat(cacheId, oidcol, key1, key2) \
+	GetSysCacheOid2(cacheId, key1, key2)
+#define GetSysCacheOid3Compat(cacheId, oidcol, key1, key2, key3) \
+	GetSysCacheOid3(cacheId, key1, key2, key3)
+#define GetSysCacheOid4Compat(cacheId, oidcol, key1, key2, key3, key4) \
+	GetSysCacheOid4(cacheId, key1, key2, key3, key4)
+
+#define LOCAL_FCINFO(name, nargs) \
+	FunctionCallInfoData name ## data; \
+	FunctionCallInfoData *name = &name ## data
+#define make_aggCompat(tlist, qual, aggstrategy, aggsplit, numGroupCols, grpColIdx, \
+					   grpOperators, grpCollations, groupingSets, chain, dNumGrous, \
+					   lefttree) \
+	make_agg(tlist, qual, aggstrategy, aggsplit, numGroupCols, grpColIdx, grpOperators, \
+			 groupingSets, chain, dNumGrous, lefttree)
+
+
+#define fcSetArg(fc, n, value) \
+	(((fc)->argnull[n] = false), ((fc)->arg[n] = (value)))
+#define fcSetArgNull(fc, n) \
+	(((fc)->argnull[n] = true), ((fc)->arg[n] = (Datum) 0))
+
+typedef struct
+{
+	File fd;
+} FileCompat;
+
+static inline int
+FileWriteCompat(FileCompat *file, char *buffer, int amount, uint32 wait_event_info)
+{
+	return FileWrite(file->fd, buffer, amount, wait_event_info);
+}
+
+
+static inline int
+FileReadCompat(FileCompat *file, char *buffer, int amount, uint32 wait_event_info)
+{
+	return FileRead(file->fd, buffer, amount, wait_event_info);
+}
+
+
+static inline FileCompat
+FileCompatFromFileStart(File fileDesc)
```

Comment:
If you can move everything related to FileCompat into the same commit (e.g., remove it from this commit and move it to FileComant commit), it'd be easier for us to review.

Reply:
So think this should be done. Brought back a change that'd been deferred until collations commit & split up `version_compat.h` to be updated bit by bit. This does introduce a kind of linear dependency on the commits, but I'm fine with that, since in the end none of these commits should really be approved with out each other. Just makes it preferable if we don't mix up their order by cherry picking them in order of approval to `pg12-master`